src/core/memory.d

```diff
@@ -810,6 +810,106 @@ struct GC
     }
 }
 
+/**
+ * Pure variants of C's memory allocation functions `malloc`, `calloc`, and
+ * `realloc` that achieve purity by aborting the program on failure so
+ * they never visibly change errno.
+ *
+ * The functions may terminate the program using `onOutOfMemoryError` or
+ * `assert(0)`. These functions' purity guarantees no longer hold if
+ * the program continues execution after catching AssertError or
+ * OutOfMemoryError.
+ *
+ * A similar fail-fast variant of C's memory deallocation function `free`
+ * is not provided due to the lack of a portable and efficient way to test
+ * for failure. If purity is required `pureFree` may be used.
+ *
+ * See_Also:
+ *     $(LREF pureMalloc), $(LREF pureCalloc), $(LREF pureRealloc),
+ *     and $(LREF pureFree) which achieve purity by saving and restoring
+ *     errno.
+ *     $(LINK2 https://dlang.org/library/object/error.html, Error class documentation)
+ *     on the danger of continued execution after catching an Error.
+ */
+void* assertMalloc()(size_t size) @nogc nothrow pure @trusted
+{
+    void* ret = fakePureMalloc(size);
+    if (!ret && size)
+    {
+        version (D_BetterC) assert(0, "Memory allocation failed");
```

Comment:
Got it.

Reply:
Also it would be better to use `version (D_Exceptions)` to check whether exceptions can be thrown as LDC/GDC allow to disable them selectively (and e.g. WebAssembly might not defined `-betterC`, but still not support exceptions.)